libcudacxx/include/cuda/std/detail/libcxx/include/support/atomic/atomic_cuda_derived.h

```diff
@@ -40,6 +44,8 @@ void _LIBCUDACXX_DEVICE __atomic_exchange_cuda(_Type volatile *__ptr, _Type *__v
 
 template<class _Type, class _Delta, class _Scope, typename _CUDA_VSTD::enable_if<sizeof(_Type)<=2, int>::type = 0>
 _Type _LIBCUDACXX_DEVICE __atomic_fetch_add_cuda(_Type volatile *__ptr, _Delta __val, int __memorder, _Scope __s) {
+    _Type __ret;
+    if (__cuda_fetch_add_weak_if_local(__ptr, __val, &__ret)) return __ret;
```

Comment:
important: compiler is unable to see through the memory and identify that it's not local. This affects codegen and overall performance. Here's a simple kernel:

```cpp
using device_atomic_t = cuda::atomic<int, cuda::thread_scope_device>;

__global__ void use(device_atomic_t *d_atomics) {
  d_atomics->fetch_add(threadIdx.x, cuda::memory_order_relaxed);
}
```

On RTX 6000 Ada the change leads to the following slowdown (up to ~3x slower)

![device_scope_atomics](https://github.com/NVIDIA/cccl/assets/9890394/a81fad95-7a7a-4608-84d9-1896ed96d8d3)

In the case of the block-scope atomics the performance difference is even more pronounced:

```cpp
template <int BlockSize>
__launch_bounds__(BlockSize) __global__ void use(device_atomic_t *d_atomics, int mv) {
  __shared__ block_atomic_t b_atomics;

  if (threadIdx.x == 0) {
    new (&b_atomics) block_atomic_t{};
  }
  __syncthreads();

  b_atomics.fetch_add(threadIdx.x, cuda::memory_order_relaxed);
  __syncthreads();

  if (threadIdx.x == 0) {
    if (b_atomics.load(cuda::memory_order_relaxed) > mv) {
      d_atomics->fetch_add(1, cuda::memory_order_relaxed);
    }
  }
}
```

Results for RTX 6000 Ada illustrate up to ~4x slowdown:

![block_scope_atomics](https://github.com/NVIDIA/cccl/assets/9890394/5ab635aa-0fc0-4c8b-8085-2bdd62af4bb1)

I think I agree with:

> Since this only impact objects with automatic storage, the impact is not very widespread

Given this, I think we should explore options not to penalize widespread use cases. If compiler is able to see through the local space check, this would be a solution. Otherwise, we can consider refining the:

> it affects an object in GPU memory and only GPU threads access it.

requirement to talk about global, cluster or block memory + add a check of automatic storage in debug build.

Reply:
This is known but the analysis is incomplete since:
- this lands on CUDA CTK 12.4,
- the impact is zero on CUDA CTK 12.3+, and
- users of CUDA CTK 12.2- using cuda atomic from their CTK are not impacted by this.

We could - in a subsequent PR - provide a way to opt out into broken behavior via some feature macro, for those users of cuda atomic from github while stuck on older CTKs, as an opt in.

Reply:
Is this something where we could work with attribute e.g `[[likely]]` / `[[unlikely]]`?